abci/types/types.go

```diff
@@ -167,3 +170,125 @@ func RespondVerifyVoteExtension(ok bool) ResponseVerifyVoteExtension {
 		Result: result,
 	}
 }
+
+func MustHashResults(r []*ExecTxResult) []byte {
+	return merkle.HashFromByteSlices(mustResultsToByteSlices(r))
+}
+
+// ProveResult returns a merkle proof of one result from the set
+func MustProveResult(r []*ExecTxResult, i int) merkle.Proof {
+	_, proofs := merkle.ProofsFromByteSlices(mustResultsToByteSlices(r))
```

Comment:
Do we need to marshal the entire result, or can we just marshal the selected one?

```go
_, proof := merkle.ProofsFromByteSlices(mustResultsToByteSlices(r[i:i+1]))
return *proof[0]
```

Relatedly: If the function returns a pointer to its result type, why are we indirecting it back off?

Reply:
> Do we need to marshal the entire result, or can we just marshal the selected one?

We need the whole result. We're building a Merkle proof inclusion in the entire list, which requires building up the hashed data structure.

I removed the function here and split out the functionality to 1. Creating a slice of []byte{} and 2. calling the merklizeing function directly. This ends up being a bit more clear and, imo, makes nicer use of the library functions instead of wrapping everything in convenience functions.